ipa-core/src/net/server/handlers/echo.rs

```diff
@@ -1,10 +1,29 @@
-use axum::{routing::get, Json, Router};
+use std::collections::HashMap;
 
-use crate::net::{http_serde, server::Error};
+use axum::{extract::Query, routing::get, Json, Router};
+use hyper::HeaderMap;
+
+use crate::net::{
+    http_serde::{self, echo::Request},
+    server::Error,
+};
 
 #[allow(clippy::unused_async)] // needs to be async for axum handler
-async fn handler(req: http_serde::echo::Request) -> Result<Json<http_serde::echo::Request>, Error> {
-    Ok(Json(req))
+async fn handler(
+    Query(query_params): Query<HashMap<String, String>>,
+    hyper_headers: HeaderMap,
+) -> Result<Json<http_serde::echo::Request>, Error> {
+    let headers = hyper_headers
+        .iter()
+        .filter_map(|(name, value)| match value.to_str() {
+            Ok(header_value) => Some((name.to_string(), header_value.to_string())),
+            Err(_) => None,
```

Comment:
you can also use `ok` method on `Result` if you don't care about errors

```suggestion
        .filter_map(|(name, value)| value.to_str().ok().map(|name| (name.to_string(), value.to_string())))
```

Reply:
Will use `ok`